File: python-algorithm/ds_4_linkedlist/linkedlist_2_inplace_operation/L92-m.py
#--coding:utf-8--


# back cur fut 分别表示上一个节点 当前节点 下一个节点

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        break_point = dummy
        cur = dummy
        cnt = 0
        while cnt < left:
            break_point = cur  # 标记反转起始位置上一个点
            cur = cur.next
            cnt += 1

        old_start = cur        # 标记反转起始位置
        back = fut = None
        while cnt <= right:
            fut = cur.next
            cur.next = back
            back = cur
            cur = fut
            cnt += 1

        break_point.next = back
        old_start.next = fut

        return dummy.next

File: ds_4_linkedlist/linkedlist_2_inplace_operation/L92-m.cpp


//部分反转链表

#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* start = dummy;
        for(int i=1 ;i<m;i++)
            start = start->next;
        //start是反转位置m的上一个节点
        ListNode* pre = NULL;
        ListNode* tem = start->next; //标记反转位置m
        ListNode* cur = start->next;
        ListNode* back = cur;
        int count = m;
        while(count <= n){
            count++;
            //先记下当前节点cur后面的节点,这里back指的是cur右边， pre是cur的左边
            back = cur->next;
            cur->next = pre;
            pre = cur;
            cur = back;
        }
        //上面循环完之后cur指向反转位置n的下一位;下面连接链表
        start->next = pre;
        tem->next = back;

        ListNode* re = dummy->next;
        delete dummy;
        return re;
    }
};